compiler/src/iree/compiler/Codegen/Common/TransformDialectInterpreterPass.cpp

```diff
@@ -0,0 +1,175 @@
+        registry);
+    tensor::registerBufferizableOpInterfaceExternalModels(registry);
+    vector::registerBufferizableOpInterfaceExternalModels(registry);
+  }
+
+  StringRef getArgument() const override {
+    return "iree-transform-dialect-interpreter";
+  }
+
+  StringRef getDescription() const override {
+    return "apply transform dialect operations one by one";
+  }
+
+  bool canScheduleOn(RegisteredOperationName name) const override {
+    return true;
+  }
+
+  TransformDialectInterpreter(StringRef transformFileName = StringRef()) {
+    this->transformFileName = transformFileName.str();
+  }
+  TransformDialectInterpreter(const TransformDialectInterpreter &pass) {
+    this->transformFileName = pass.transformFileName;
+    // TODO: if we really don't like shared_ptr, we could also clone the
+    // transformModule here.
+    sharedTransformModule = pass.sharedTransformModule;
+  }
+
+  LogicalResult initialize(MLIRContext *context) override {
+    OwningOpRef<ModuleOp> module;
+    if (failed(transform::parseTransformModuleFromFile(
+            context, transformFileName, module)))
+      return failure();
+    sharedTransformModule =
+        std::make_shared<OwningOpRef<ModuleOp>>(std::move(module));
+    return success();
+  }
+
+  void runOnOperation() override {
+    Operation *target = getOperation();
+    bool parsedTransform = (sharedTransformModule && *sharedTransformModule);
+    assert(parsedTransform || (target->getNumRegions() == 1 &&
+                               target->getRegion(0).getBlocks().size() == 1) &&
+                                  "Cannot extract transform from op");
+    Region &transformRegion = parsedTransform
+                                  ? (*sharedTransformModule)->getRegion()
+                                  : target->getRegion(0);
+    if (failed(transform::applyTransformsInRegion(transformRegion, target)))
+      return signalPassFailure();
+  }
+
+ protected:
+  Pass::Option<std::string> transformFileName{
+      *this, "transform-file-name",
+      ::llvm::cl::desc(
+          "File name containing a transform dialect specification to apply.")};
+
+ private:
+  // The parsed transform module to be used for transformations.
+  // TODO: Figure a better way to build a transform module and transport it in
+  // the proper places in the IR as it is transformed by IREE so that it is
+  // available with better ownership semantics.
+  // Note: we wrap the OwningOpRef to get the desired destruction mechanism.
+  // Note: shared_ptr is not great but we know the sharedTransformModule is
+  // readonly.
+  // Alternatives comprise:
+  //   1. no shared_ptr but copying the module with every pass clone that the
+  //      OpPassManager decides to perform.
+  //   2. lifting ownership of the parsed transform module higher up in the
+  //      IREE stack. This may be only shift the problem as we have passes
+  //      building pass managers in IREE.
```

Comment:
Do we? We have dynamic pass pipelines but should not have nested pass managers.

Reply:
I think things related to bufferization do, e.g. https://sourcegraph.com/github.com/iree-org/iree/-/blob/compiler/src/iree/compiler/Codegen/Common/BufferizeCopyOnlyDispatchesPass.cpp?L82

The sandbox / codegen strategies stuff also has that for lack of better control pre-transform dialect.

I thought I had seen more but looking at the LLVMCPU passes I see the pass manager is properly passed around and populated.

So it may just be some cleanups and we may be able to revisit the ownership model (but that would come with more intertwined IREE / transform dialect which may not be what we want. Am despite the include the world behavior due to registration needs, things are quite isolated logically).

Reply:
I also had some older TODO that I am not 100% sure about, it was complaining about lowering to LLVM using its own pass manager and me having had unhappy experiences with registration in general.